include/CObjectCache.h

```diff
@@ -9,58 +9,110 @@
-        using NoPtrValueType = typename std::remove_pointer<ValueType>::type;
-        using PtrToConstVal_t = const NoPtrValueType*; // ValueType is always pointer type
+        // typedefs for implementation only
+        using PairType_impl = typename ContainerT::value_type;
+        //! Always pointer type
+        using ValueType_impl = typename PairType_impl::second_type; // container's value_type is always instantiation of std::pair
+        static_assert(std::is_pointer<ValueType_impl>::value, "ValueType_impl must be pointer type!");
+        using KeyType_impl = typename PairType_impl::first_type;
+        using NoPtrValueType_impl = typename std::remove_pointer<ValueType_impl>::type;
+        using ValueType_PtrToConst_impl = const NoPtrValueType_impl*; // ValueType_impl is always pointer type
+
+    public:
+        using RangeType = std::pair<IteratorType, IteratorType>;
+        using ConstRangeType = std::pair<ConstIteratorType, ConstIteratorType>;
+
+        using GreetFuncType = std::function<void(ValueType_impl)>;
+        using DisposalFuncType = std::function<void(ValueType_impl)>;
+
+        template<typename RangeT>
+        static bool isNonZeroRange(const RangeT& _range) { return std::distance(_range.first, _range.second) != 0; }
 
-        std::function<void(ValueType)> m_greetingFunc, m_disposalFunc;
+    protected:
+        GreetFuncType m_greetingFunc;
+        DisposalFuncType m_disposalFunc;
     protected:
 		inline virtual ~CObjectCacheBase()
 		{
 			for (auto it=m_container.begin(); it!=m_container.end(); it++)
 				dispose(it->second);
 		}
 
-        void dispose(ValueType _object) const        {
+        void dispose(ValueType_impl _object) const        {
             if (m_disposalFunc)
                 m_disposalFunc(_object);
         }
 
-        void greet(ValueType _object) const
+        void greet(ValueType_impl _object) const
         {
             if (m_greetingFunc)
                 m_greetingFunc(_object);
         }
 
     public:
-        inline explicit CObjectCacheBase(const std::function<void(ValueType)>& _greeting, const std::function<void(ValueType)>& _disposal) : m_greetingFunc(_greeting), m_disposalFunc(_disposal) {}
-        inline explicit CObjectCacheBase(std::function<void(ValueType)>&& _greeting, std::function<void(ValueType)>&& _disposal) : m_greetingFunc(std::move(_greeting)), m_disposalFunc(std::move(_disposal)) {}
+        inline explicit CObjectCacheBase(const GreetFuncType& _greeting, const DisposalFuncType& _disposal) : m_greetingFunc(_greeting), m_disposalFunc(_disposal) {}
+        inline explicit CObjectCacheBase(GreetFuncType&& _greeting, DisposalFuncType&& _disposal) : m_greetingFunc(std::move(_greeting)), m_disposalFunc(std::move(_disposal)) {}
 
-        inline bool contains(PtrToConstVal_t _object) const
+        inline bool contains(ValueType_PtrToConst_impl _object) const
```

Comment:
No, it'd be `T* const` and I want `const T*`

Reply:
Granted, we could do the following as a replacement to `ValueType_PtrToConst_impl`
```
constexpr static isTypeAPointer = std::is_pointer<ValueType_impl>::value;
typename std::conditional<isTypeAPointer,const typename std::remove_pointer<ValueType_impl>::type*,const ValueType_impl&>::type
```

This would leverage SFINAE to handle non-pointer cached object types

P.S. Right now non-ptr objects will fail because they won't be able to instantiate the `NoPtrValueType_impl` and `ValueType_PtrToConst_impl` aliases.

Reply:
Non-pointers will fail because that's how cache is designed and it's also made sure by `static_assert`. Also I'm not sure if you understand how it is. Cache's template parameter takes any type, it doesn't have to be pointer type! But underlaying container **always** contains pointer type (and so `ValueType_impl` typedef is always pointer type) which is quite obvious since it's a cache and not a regular container and i believe that's why you initially asked for cache as container for pointers.
Example code:
```
core::CConcurrentObjectCache<core::ICPUBuffer, std::string, std::unordered_map> cache;
core::ICPUBuffer* b = new core::ICPUBuffer(0xff);
cache.insert("b", b);
cache.removeObject(b, "b");
```

Reply:
damn...

I thought I would instantiate the cache as `core::CConcurrentObjectCache<core::ICPUBuffer*, std::string, std::unordered_map>`


> i believe that's why you initially asked for cache as container for pointers.

Its not required for the cache to work on all types, but I thought if it could be achieved through changing a single line of code, then it should be done.

Reply:
Also that's why there's two typedefs: `CachedType` (public) and `ValueType_impl` (private)